llvm/labs/lab2/ulyanov_daniil/mulToBitShift.cpp

```diff
@@ -0,0 +1,70 @@
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+
+using namespace llvm;
+
+namespace {
+class MTBSPass : public PassInfoMixin<MTBSPass> {
+public:
+  PreservedAnalyses run(llvm::Function &F, llvm::FunctionAnalysisManager &AM) {
+    for (auto &BB : F) {
+      for (auto InstIt = BB.begin(); InstIt != BB.end(); InstIt++) {
+        if (InstIt->getOpcode() != Instruction::Mul) {
+          continue;
+        }
+
+        auto *Op = dyn_cast<BinaryOperator>(InstIt);
+        Value *LVal = Op->getOperand(0);
+        Value *RVal = Op->getOperand(1);
+
+        int LLog = getLog2(LVal);
+        int RLog = getLog2(RVal);
+
+        if (RLog < LLog) {
+          std::swap(LLog, RLog);
+          std::swap(LVal, RVal);
+        }
+        if (RLog < 0) {
+          continue;
+        }
+
+        IRBuilder<> Builder(Op);
+        Value *NewVal =
```

Comment:
Please align shift step on a size of shifted type. I.e. if shifted value type is int32 and Rlog > 32 then set it just to 32. The same logic is for other types (int8/16/23/...).